[JSC] Use destroying delete for ArrayBufferView

Constellation opened the pull request:
#### 4f3aa596bc546e675bd4cc421b1e125051ec7352
<pre>
[JSC] Use destroying delete for ArrayBufferView
<a href="https://bugs.webkit.org/show_bug.cgi?id=245653">https://bugs.webkit.org/show_bug.cgi?id=245653</a>

Reviewed by Darin Adler.

Set TypedArrayType in ArrayBufferView field and use destroying delete for destruction and the other tiny functions.
So, we remove vtable pointer for that class.

* Source/JavaScriptCore/runtime/ArrayBufferView.cpp:
(JSC::ArrayBufferView::ArrayBufferView):
(JSC::ArrayBufferView::visitDerived):
(JSC::ArrayBufferView::visitDerived const):
(JSC::ArrayBufferView::wrap):
(JSC::ArrayBufferView::operator delete):
* Source/JavaScriptCore/runtime/ArrayBufferView.h:
(JSC::ArrayBufferView::getType const):
* Source/JavaScriptCore/runtime/DataView.cpp:
(JSC::DataView::DataView):
(JSC::DataView::wrapImpl):
(JSC::DataView::wrap): Deleted.
* Source/JavaScriptCore/runtime/DataView.h:
* Source/JavaScriptCore/runtime/GenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h:
(JSC::GenericTypedArrayView&lt;Adaptor&gt;::GenericTypedArrayView):
(JSC::GenericTypedArrayView&lt;Adaptor&gt;::wrapImpl):
(JSC::GenericTypedArrayView&lt;Adaptor&gt;::wrap): Deleted.
* Source/JavaScriptCore/runtime/TypedArrayType.h:

Canonical link: <a href="https://commits.webkit.org/254891@main">https://commits.webkit.org/254891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d25f79a857b8d05b791fc1361e33b3de4599a22b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90571 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35155 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21206 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99904 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94581 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33655 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82935 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96334 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96226 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77416 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26629 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81566 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81365 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69652 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82144 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34754 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15402 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77093 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32569 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/16380 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26620 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3419 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36334 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79679 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35469 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17461 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/ArrayBufferView.cpp:
```diff
@@ -41,10 +44,41 @@ ArrayBufferView::ArrayBufferView(
+    switch (m_type) {
+    case TypedArrayType::NotTypedArray:
+    case TypedArrayType::TypeDataView:
+        return std::invoke(std::forward<Visitor>(visitor), *static_cast<DataView*>(this));
```
Minor stylistic thing, I would write `static_cast<DataView&>(*this)` instead.

Constellation replied:
Thanks, changed.

darinadler commented on Source/JavaScriptCore/runtime/ArrayBufferView.cpp:
```diff
@@ -41,10 +44,41 @@ ArrayBufferView::ArrayBufferView(
+void ArrayBufferView::operator delete(ArrayBufferView* value, std::destroying_delete_t)
+{
+    value->visitDerived([](auto& value) {
+        using T = std::decay_t<decltype(value)>;
```
Seems to me that putting this all into the `freeAfterDestruction` line would be just fine without a named type: `std::decay_t<decltype(value)>::freeAfterDestruction(&value)`, unless maybe that doesn’t compile for some reason?

Constellation replied:
I think this is better, here is the reason. Currently, this class is just a RefCounted. So we do not need to call a specific free function. But for example ISO_ALLOCATED ones need to call specific free functions, and this way can call appropriate free functions defined per class. And at the same time, if they are the same, since `freeAfterDestruction` is inline function, we expect that the generated code will be the same to call it outside of this `visitDerived`. :)
So, this way can prevent introducing a bug when ISO_ALLOCATED is used, while this way does not regress things for the other cases, so I think just using this style for all would be simpler rule.

darinadler replied:
You may have misunderstood my feedback. I was suggesting a different format for the code where it is all on one line, not omitting the use of `std::decay_t<decltype(value)>` entirely. Further, I would encourage you to add the `std::decay_t<decltype(value)>` idiom to existing destroying delete implementations that should have it (maybe just the one).

Constellation replied:
Ah! I see. Thanks, that makes sense

darinadler commented on Source/JavaScriptCore/runtime/ArrayBufferView.h:
```diff
@@ -109,11 +109,16 @@ class ArrayBufferView : public RefCounted<ArrayBufferView> {
 protected:
-    JS_EXPORT_PRIVATE ArrayBufferView(RefPtr<ArrayBuffer>&&, size_t byteOffset, size_t byteLength);
+    template<typename Visitor> constexpr decltype(auto) visitDerived(Visitor&&);
+    template<typename Visitor> constexpr decltype(auto) visitDerived(Visitor&&) const;
```
Seems like these can be private, not just protected.

Constellation replied:
Changed.

darinadler commented on Source/JavaScriptCore/runtime/DataView.h:
```diff
@@ -34,13 +34,8 @@ class DataView final : public ArrayBufferView {
-    }
 
-    JSArrayBufferView* wrap(JSGlobalObject*, JSGlobalObject*) final;
+    JSArrayBufferView* wrapImpl(JSGlobalObject*, JSGlobalObject*);
```
I’m personally not all that big a fan of the "Impl" suffix as a way to name things like this, but it’s consistent with use elsewhere. Also seems to me that the two arguments need names. It’s not obvious which means which when they are both arguments of the same type.

Constellation replied:
Added parameter names. For now, as a naming convention aligned with the other places, using `wrapImpl`.

darinadler commented on Source/JavaScriptCore/runtime/ArrayBufferView.cpp:
```diff
@@ -41,10 +44,41 @@ ArrayBufferView::ArrayBufferView(
+    FOR_EACH_TYPED_ARRAY_TYPE_EXCLUDING_DATA_VIEW(DECLARE_TYPED_ARRAY_TYPE)
+#undef DECLARE_TYPED_ARRAY_TYPE
+
+    }
```
I’d omit this blank line.

Constellation replied:
Removed

darinadler commented on Source/JavaScriptCore/runtime/ArrayBufferView.cpp:
```diff
@@ -41,10 +44,41 @@ ArrayBufferView::ArrayBufferView(
+    case TypedArrayType::TypeDataView:
+        return std::invoke(std::forward<Visitor>(visitor), *static_cast<DataView*>(this));
+#define DECLARE_TYPED_ARRAY_TYPE(name) \
+    case TypedArrayType::Type ## name: \
```
It’s a little surprising to have spaces around the `##`.

Constellation replied:
Changed.

darinadler commented on Source/JavaScriptCore/runtime/ArrayBufferView.h:
```diff
@@ -109,11 +109,16 @@ class ArrayBufferView : public RefCounted<ArrayBufferView> {
-    virtual JSArrayBufferView* wrap(JSGlobalObject*, JSGlobalObject*) = 0;
-    
+
+    JS_EXPORT_PRIVATE JSArrayBufferView* wrap(JSGlobalObject*, JSGlobalObject*);
```
I think these arguments need names.

Constellation replied:
Added.

Constellation reviewed (commented):
Thanks!

Constellation committed:
[JSC] Use destroying delete for ArrayBufferView
https://bugs.webkit.org/show_bug.cgi?id=245653

Reviewed by Darin Adler.

Set TypedArrayType in ArrayBufferView field and use destroying delete for destruction and the other tiny functions.
So, we remove vtable pointer for that class.

* Source/JavaScriptCore/runtime/ArrayBufferView.cpp:
(JSC::ArrayBufferView::ArrayBufferView):
(JSC::ArrayBufferView::visitDerived):
(JSC::ArrayBufferView::visitDerived const):
(JSC::ArrayBufferView::wrap):
(JSC::ArrayBufferView::operator delete):
* Source/JavaScriptCore/runtime/ArrayBufferView.h:
(JSC::ArrayBufferView::getType const):
* Source/JavaScriptCore/runtime/DataView.cpp:
(JSC::DataView::DataView):
(JSC::DataView::wrapImpl):
(JSC::DataView::wrap): Deleted.
* Source/JavaScriptCore/runtime/DataView.h:
* Source/JavaScriptCore/runtime/GenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h:
(JSC::GenericTypedArrayView<Adaptor>::GenericTypedArrayView):
(JSC::GenericTypedArrayView<Adaptor>::wrapImpl):
(JSC::GenericTypedArrayView<Adaptor>::wrap): Deleted.
* Source/JavaScriptCore/runtime/TypedArrayType.h:

Canonical link: https://commits.webkit.org/254891@main
